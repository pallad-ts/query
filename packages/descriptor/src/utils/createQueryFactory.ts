import { z } from "zod";
import { fromNullable } from "@sweet-monads/maybe";
import { QueryDescriptor } from "../QueryDescriptor";
import { fromZodValidation } from "@pallad/violations-zod";
import { right } from "@sweet-monads/either";

export function createQueryFactory(
	config: QueryDescriptor.Config,
	filtersValidator: QueryDescriptor.FiltersValidator<any> | undefined
) {
	const paginationSchema = getPaginationSchema(config.pagination);
	const sortingSchema = getSortingSchema(config.sorting);

	let schema = z.object({
		filters: z
			.object({})
			.passthrough()
			.default(() => ({})),
	});

	schema = paginationSchema.fold(
		() => schema,
		extraSchema => schema.merge(extraSchema)
	);
	schema = sortingSchema.fold(
		() => schema,
		extraSchema => schema.merge(extraSchema)
	);

	return (input: unknown) => {
		return fromZodValidation(schema, input).chain(({ filters, ...meta }) => {
			if (filtersValidator) {
				return filtersValidator(filters).map(filters => ({ ...meta, filters }));
			}
			return right({ ...meta, filters });
		});
	};
}

function createSortableFieldSchema(allowedFields: readonly [string, ...string[]]) {
	return z.object({
		field: z.enum(allowedFields),
		direction: z.enum(["DESC", "ASC"]),
	});
}

function getPaginationSchema(pagination: QueryDescriptor.Pagination | undefined) {
	return fromNullable(pagination).map(x => {
		if (x.type === "byCursor") {
			return z.object({
				limit: getLimitSchema(x),
				after: z.string().optional(),
				before: z.string().optional(),
			});
		}
		return z.object({
			limit: getLimitSchema(x),
			offset: z.number().int().min(0).default(0),
		});
	});
}

function getLimitSchema(pagination: QueryDescriptor.Pagination) {
	return z.number().int().min(1).max(pagination.maxLimit).default(pagination.defaultLimit);
}

function getSortingSchema(sorting: QueryDescriptor.Sorting<any> | undefined) {
	return fromNullable(sorting).map(x => {
		const sortByType = createSortableFieldSchema(x.sortableFields);
		if (x.type === "single") {
			return z.object({
				sortBy: sortByType.default(x.defaultSorting),
			});
		}

		return z.object({
			sortBy: z
				.array(sortByType)
				.nonempty()
				.default(x.defaultSorting as [any, ...any[]]),
		});
	});
}
