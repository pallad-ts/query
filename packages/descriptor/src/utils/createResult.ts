import { fromNullable, Maybe } from "@sweet-monads/maybe";
import {
	PaginableByCursor,
	PaginableByOffset,
	Query,
	SortableMulti,
	SortableSingle,
} from "@pallad/query";
import { QueryDescriptor } from "../QueryDescriptor";

export function createResult(
	config: QueryDescriptor.Config,
	resultOptions: QueryDescriptor.ResultOptions<any, any, any>
) {
	let meta: any = {};
	composeMetaForPagination(config.pagination, resultOptions).chain(x =>
		Object.assign(meta, x)
	);
	composeMetaForSorting(config.sorting, resultOptions.query).chain(x => Object.assign(meta, x));

	const hasMeta = Object.keys(meta).length > 0;

	return {
		results: resultOptions.results,
		...(hasMeta ? { meta } : {}),
	};
}

function composeMetaForPagination(
	pagination: QueryDescriptor.Pagination | undefined,
	options: QueryDescriptor.ResultOptions<any, any, any>
) {
	return fromNullable(pagination).map(settings => {
		if (settings.type === "byOffset") {
			const finalQuery = options.query as Required<PaginableByOffset>;
			return {
				limit: finalQuery.limit,
				offset: finalQuery.offset,
			} as PaginableByOffset.ResultMeta;
		}

		const finalQuery = options.query as Required<PaginableByCursor>;
		return {
			limit: finalQuery.limit,
			nextPage: "nextPageCursor" in options ? options.nextPageCursor : undefined,
			previousPage: "previousPageCursor" in options ? options.previousPageCursor : undefined,
		} as PaginableByCursor.ResultMeta;
	});
}

function composeMetaForSorting(
	sorting: QueryDescriptor.Sorting<any> | undefined,
	query: Query<any> & (SortableMulti<any> | SortableSingle<any>)
) {
	return fromNullable(sorting).map(() => {
		return { sortBy: query.sortBy } as
			| SortableSingle.ResultMeta<any>
			| SortableMulti.ResultMeta<any>;
	});
}
