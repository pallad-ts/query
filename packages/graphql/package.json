{
  "name": "@pallad/query-graphql",
  "version": "2.0.1",
  "description": "Helper for graphql to create @pallad/query related types",
  "main": "compiled/index.js",
  "types": "compiled/index.d.ts",
  "scripts": {
    "test": "pallad-scripts test",
    "lint": "pallad-scripts lint",
    "compile": "pallad-scripts compile",
    "prepublishOnly": "npm run compile"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/pallad-ts/query.git",
    "directory": "packages/main"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Łukasz Kużyński <lukasz.kuzynski@gmail.com> (http://wookieb.pl/)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pallad-ts/query/issues"
  },
  "homepage": "https://github.com/pallad-ts/query#readme",
  "peerDependencies": {
    "graphql": "^16.3.0",
    "graphql-compose": "^9.0.8"
  },
  "dependencies": {
    "@pallad/builder": "^1.0.1",
    "@pallad/query": "^2.0.0",
    "predicates": "^2.0.3"
  }
}
