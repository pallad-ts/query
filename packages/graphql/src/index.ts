export * from './types';
export * from './createInputSortType';
export * from './createResultListType';
export * from './createSortFieldType';
export * from './createSortTypeBase';
export * from './GraphQLQueryBuilder';
