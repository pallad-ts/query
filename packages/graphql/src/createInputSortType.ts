import {GraphQLEnumType} from "graphql";
import {createSortTypeBase} from "./createSortTypeBase";

/**
 * Creates `sort` type suitable for use as input
 */
export function createInputSortType({baseName, sortFieldType, isMulti}: createInputSortType.Options) {
	const type = createSortTypeBase(`${baseName}_Query_Sort`, sortFieldType)
		.getInputTypeComposer({postfix: '_Input'});

	if (isMulti) {
		return type.NonNull.List;
	}
	return type;
}

export namespace createInputSortType {
	export interface Options {
		baseName: string,
		sortFieldType: GraphQLEnumType,
		/**
		 * Whether multiple sorting fields are allowed
		 */
		isMulti: boolean;
	}
}

