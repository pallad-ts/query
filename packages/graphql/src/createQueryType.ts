import {InputTypeComposerFieldConfigMapDefinition} from "graphql-compose/lib/InputTypeComposer";
import {ComposeInputType} from "graphql-compose/lib/utils/typeHelpers";
import {InputTypeComposer} from "graphql-compose";
import {GraphQLInputObjectType, GraphQLNonNull} from "graphql";

export function createQueryType<TContext>({baseName, filtersType, paginationFields, sortType}: createQueryType.Options<TContext>) {
	const wrappedFiltersType = filtersType instanceof GraphQLInputObjectType ? new GraphQLNonNull(filtersType) : filtersType.NonNull;
	return InputTypeComposer.createTemp<TContext>({
		name: `${baseName}_Query`,
		fields: {
			filters: {type: wrappedFiltersType},
			...(sortType ? {
				sortBy: {
					type: sortType
				}
			} : undefined),
			...paginationFields
		}
	});
}

export namespace createQueryType {
	export interface Options<TContext> {
		baseName: string;
		filtersType: GraphQLInputObjectType | InputTypeComposer<TContext>,
		paginationFields?: InputTypeComposerFieldConfigMapDefinition;
		sortType?: ComposeInputType;
	}
}
