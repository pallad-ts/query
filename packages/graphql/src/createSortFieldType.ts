import {GraphQLEnumType} from "graphql";
import * as is from 'predicates'
import {camelCase} from "camel-case";

const assertSortableFields = is.assert(is.notEmptyArr);

/**
 * Creates enum for sortable fields
 */
export function createSortFieldType(baseName: string, sortableFields: readonly string[]) {
	assertSortableFields(sortableFields);
	return new GraphQLEnumType({
		name: `${baseName}_Sort_Field`,
		values: Object.fromEntries(
			sortableFields.map(x => {
				return [camelCase(x), {value: x}]
			})
		)
	})
}
