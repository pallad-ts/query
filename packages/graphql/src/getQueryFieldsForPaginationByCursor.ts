import { GraphQLString } from "graphql";
import { GraphQLPositiveInt } from "graphql-scalars";
import { InputTypeComposerFieldConfigMapDefinition } from "graphql-compose/lib/InputTypeComposer";

export function getQueryFieldsForPaginationByCursor(): InputTypeComposerFieldConfigMapDefinition {
	return {
		after: { type: GraphQLString },
		before: { type: GraphQLString },
		limit: { type: GraphQLPositiveInt },
	};
}
