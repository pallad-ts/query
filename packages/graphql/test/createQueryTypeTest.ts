import {createQueryType} from "@src/createQueryType";
import {GraphQLInputObjectType, GraphQLNonNull, GraphQLString} from "graphql";
import {createInputSortType} from "@src/createInputSortType";
import {createSortFieldType} from "@src/createSortFieldType";
import {SORTABLE_FIELDS} from "./fixtures";
import {getQueryFieldsForPaginationByCursor} from "@src/getQueryFieldsForPaginationByCursor";
import {printInputObject} from "graphql-compose/lib/utils/schemaPrinter";

describe('createQueryType', () => {
	const filtersType = new GraphQLInputObjectType({
		name: 'Base_Filters',
		fields: {
			name: {type: new GraphQLNonNull(GraphQLString)}
		}
	});

	const sortType = createInputSortType({
		baseName: 'Base',
		sortFieldType: createSortFieldType('Base', SORTABLE_FIELDS),
		isMulti: false
	});

	it.each<[string, Omit<createQueryType.Options<any>, 'baseName'>]>([
		[
			'simple',
			{
				filtersType,
			}
		],
		[
			'with sort type',
			{
				filtersType,
				sortType
			}
		],
		[
			'with pagination fields',
			{
				filtersType,
				paginationFields: getQueryFieldsForPaginationByCursor()
			}
		],
		[
			'with all options',
			{
				filtersType,
				sortType,
				paginationFields: getQueryFieldsForPaginationByCursor()
			}
		]
	])('created query input type', (_, opts) => {
		expect(
			printInputObject(
				createQueryType({
					baseName: 'Base',
					...opts
				}).getType()
			)
		)
			.toMatchSnapshot();
	})
});
