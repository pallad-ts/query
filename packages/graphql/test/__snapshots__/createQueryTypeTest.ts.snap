// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createQueryType created query input type 1`] = `
"input Base_Query {
  filters: Base_Filters!
}"
`;

exports[`createQueryType created query input type 2`] = `
"input Base_Query {
  filters: Base_Filters!
  sortBy: Base_Query_Sort_Input
}"
`;

exports[`createQueryType created query input type 3`] = `
"input Base_Query {
  filters: Base_Filters!
  after: String
  before: String
  limit: PositiveInt
}"
`;

exports[`createQueryType created query input type 4`] = `
"input Base_Query {
  filters: Base_Filters!
  sortBy: Base_Query_Sort_Input
  after: String
  before: String
  limit: PositiveInt
}"
`;
